privacyidea/lib/importotp.py

```diff
@@ -519,18 +527,43 @@ def parsePSKCdata(xml_data,
                                           "AES128-CBC.")
                 enc_data = key.data.secret.encryptedvalue.ciphervalue.text
                 enc_data = enc_data.strip()
-                secret = aes_decrypt_b64(binascii.unhexlify(preshared_key_hex), enc_data)
+
+                preshared_key = binascii.unhexlify(preshared_key_hex)
+
+                secret = aes_decrypt_b64(preshared_key, enc_data)
+
                 if token["type"].lower() in ["hotp", "totp"]:
                     token["otpkey"] = hexlify_and_unicode(secret)
                 elif token["type"].lower() in ["pw"]:
                     token["otpkey"] = to_unicode(secret)
                 else:
                     token["otpkey"] = to_unicode(secret)
+
+                if validate_mac != 'no_check':
+                    # Validate MAC:
+                    encrypted_mac_key = xml.keycontainer.find("mackey").text
+                    mac_key = aes_decrypt_b64(preshared_key, encrypted_mac_key)
+
+                    enc_data_bin = base64.b64decode(enc_data)
+                    hm = hmac.new(key=mac_key, msg=enc_data_bin, digestmod=hashlib.sha1)
+                    mac_value_calculated = b64encode_and_unicode(hm.digest())
+
+                    mac_value_xml = key.data.find('valuemac').text.strip()
+
+                    is_invalid = not hmac.compare_digest(mac_value_xml, mac_value_calculated)
+
+                    if is_invalid and validate_mac == 'check_fail_hard':
+                        abort = True
```

Comment:
If you actually want to abort and leave the for-loop on an error, you can add a

~~~~python
break
~~~~
here, which will leave the for-loop immediately.

Reply:
This is done to get the list of all serials at the end, if I break the loop here I would have to parse all serials from the .xml again later.

Reply:
OK, thx.

Reply:
OK. thx